### PR TITLE
[batch] allow jobs to request a port

### PR DESCRIPTION
**Pull request by cseed**
This is for the batch callback test, so I can make test_batch a direct pod.

### CODE REVIEW

File: batch/batch/worker.py [hunk 1]
```diff
@@ -58,6 +58,31 @@
...
 
+port_allocator = None
+
+worker = None
+
+
+class PortAllocator:
+    def __init__(self):
+        # random unallocated ports take from
+        # https://stackoverflow.com/questions/10476987/best-tcp-port-number-range-for-internal-applications
```
Comment (danking) [inline batch/batch/worker.py]: I don't see how this SO post is relevant to the choice of the range 46572-46582? Is 46572 just a random choice from one of the unassigned ranges in the second answer?

File: batch/batch/worker.py [hunk 2]
```diff
@@ -58,6 +58,31 @@
...
+        while True:
+            if self.ports:
+                return self.ports.pop()
+            async with self.cond:
+                await self.cond.wait()
+
+    async def free(self, port):
+        self.ports.append(port)
+        async with self.cond:
+            self.cond.notify()
```
Comment (danking) [inline batch/batch/worker.py]: This seems like a more complicated implementation than:
```python3
class PortAllocator:
    def __init__(self):
        self.ports = asyncio.Queue()
        port_base = 46572
        for port in range(port_base, port_base + 10):
            self.ports.put_nowait(port)

    async def allocate(self):
        return await self.ports.get()

    def free(self, port):
        self.ports.put_nowait(port)
```

File: batch/batch/worker.py [hunk 3]
```diff
@@ -146,7 +174,20 @@ def container_config(self):
...
         }
 
-        env = self.spec.get('env')
+        env = self.spec.get('env', [])
+
+        if self.port is not None:
+            assert self.host_port is not None
+            config['PortBindings'] = {
+                f'{self.port}/tcp': [{
+                    'HostIp': '',
```
Comment (danking) [inline batch/batch/worker.py]: I can't find documentation on the docker API. Can you point me to the docs that explain what empty string means for HostIp?

File: batch/batch/worker.py [hunk 4]
```diff
@@ -395,13 +444,11 @@ def __init__(self, batch_id, user, gsa_key, job_spec):
         if job_spec.get('mount_docker_socket'):
             main_volume_mounts.append('/var/run/docker.sock:/var/run/docker.sock')
 
-        if pvc_size or input_files or output_files:
-            self.mount_io = True
+        self.mount_io = (pvc_size or input_files or output_files)
```
Comment (danking) [inline batch/batch/worker.py]: unnecessary parentheses

File: hail/python/hailtop/batch_client/validate.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 #   'mount_docker_socket': bool,
```
Comment (danking) [inline hail/python/hailtop/batch_client/validate.py]: Well this is one heck of a file. How m much faster is this than Cerberus?

File: batch/batch/worker.py (hunk 1 above)
Comment (cseed) [reply batch/batch/worker.py]: Yes.  I updated the link to point to the second answer.

File: batch/batch/worker.py (hunk 3 above)
Comment (cseed) [reply batch/batch/worker.py]: Of course docker isn't documented to this level of detail.  In general, the API docs are here (create container, for example): https://docs.docker.com/engine/api/v1.40/#operation/ContainerCreate

Normally, I figure out the config I want by using the docker command line to create an example container and then inspect the configuration it created with `docker inspect`.  That's where this came from.  This also seems sensible, and I believe empty string means bind all host interfaces.

**Pull request merged**